### ISSUE TITLE
MariaDB 10.2.7 BC break

### ISSUE THREAD

**Issue by bilhackmac** [doctrine/doctrine2#6565]
MariaDB 10.2.7 and future releases (5.5, 10.0 and 10.1 branches are not affected) just resolve an old bug in SQL standards implementation and introduce a BC break

Ticket that explains changes https://jira.mariadb.org/browse/MDEV-13132
Ticket that confirms (in comments) only >= v10.2.7 are affected https://jira.mariadb.org/browse/MDEV-13341

Concretely, for Doctrine, schema update just gone crazy and want to update every column with _NULL_ default because **'NULL'** is return by MariaDB instead of _NULL_.
I don't find any other issues… for now.

For the moment, MySQL don't have the same issue.

**Comment by Ocramius** [comment]
Needs a test case though...

**Comment by sagaraspl** [comment]
Hi @bilhackmac I also faced the same issue when I updated with MariaDB 10.2.7

**Comment by belgattitude** [comment]
Same issue too.

**Comment by baumannsven** [comment]
I have the same problem.

**Comment by lcobucci** [comment]
@sabzeta @belgattitude @baumannsven as @Ocramius said, please send us a PR with tests. That helps A LOT 😉

**Comment by sabzeta** [comment]
I haven't realy ran into this problem, but it's good to be aware. I'll try to look into it. 
Meanwhile, perhaps @sagaraspl could send a PR with tests?

**Comment by belgattitude** [comment]
Hi @lcobucci , @Ocramius  

It can be fixed in many ways... But before making a P/R I would like your opinion, advices... I'm not familiar with doctrine...

So here's the approaches I think of (easy fixes)

1. First approach

    Changing the way the schema is read directly in the `getListTableColumnsSQL` of 
    [DBAL/Platforms/MySqlPlatform.php](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L387)

    ```sql
    SELECT `COLUMN_NAME` AS `Field`,
       `COLUMN_TYPE` AS `Type`,
       `IS_NULLABLE` AS `Null`,
       `COLUMN_KEY` AS `Key`,
       --- !!!! REPLACE  `COLUMN_DEFAULT` AS `Default`, BY THE CODE BELOW !!!! ---
       IF (`IS_NULLABLE` = 'YES' and `COLUMN_DEFAULT` = 'NULL', null, `COLUMN_DEFAULT`) as `Default`,
       `EXTRA` AS `Extra`,
       `COLUMN_COMMENT` AS `Comment`,
       `CHARACTER_SET_NAME` AS `CharacterSet`,
       `COLLATION_NAME` AS `Collation`
    FROM `information_schema`.`COLUMNS`
    WHERE `TABLE_SCHEMA` = '$database' AND `TABLE_NAME` = '$table'
    ```
2. Second approach

    As it looks mariadb only, maybe create a specific abstraction... Just like  you did with [Mysql57Platform](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Platforms/MySQL57Platform.php) ???

3. Third approach
    
    Updating the method [_getPortableTableColumnDefinition] of [DBAL/Schema/MysqlSchemaManager](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php#L185)

    ```php
        $options = array(
            'length'        => $length,
            'unsigned'      => (bool) (strpos($tableColumn['type'], 'unsigned') !== false),
            'fixed'         => (bool) $fixed,
            // REPLACE THE LINE BELOW
            //    'default'       => isset($tableColumn['default']) ? $tableColumn['default'] : null,
            // BY THIS ONE
            'default'       => isset($tableColumn['default']) && !($tableColumn['default'] === 'NULL' && $tableColumn['null'] === 'YES') ? $tableColumn['default'] : null,
            'notnull'       => (bool) ($tableColumn['null'] != 'YES'),
            'scale'         => null,
            'precision'     => null,
            'autoincrement' => (bool) (strpos($tableColumn['extra'], 'auto_increment') !== false),
            'comment'       => isset($tableColumn['comment']) && $tableColumn['comment'] !== ''
                ? $tableColumn['comment']
                : null,
        );
    ```

Personally I prefer the third approach... I'll clean up a bit the code I show you of course (would like to set default in a separate var, so we can eventually check the server version).

I'll publish a P/R on doctrine/dbal as soon as you give me the preferred way. For the test case, I'll have to check a bit how to do it ;)

**Comment by belgattitude** [comment]
just published a P/R on https://github.com/doctrine/dbal/pull/2824 with the third approach... Feel free to comment.

**Comment by mdeboer** [comment]
Same problem here, narrowed it down to it being a NULL problem of some sort as removing all nullable relations and fields solved it. That combined with me recently pulling the newest mariadb image for Docker explains it all.

There went another 3/4 hours of my day trying to find an error in my code even though the mapping files validated fine 😭  Using the mysql:5 image now which should fix things, fingers crossed 🤞 .

Hopefully a fix will be merged soon!

**Comment by belgattitude** [comment]
@bilhackmac, @mdeboer @baumannsven @sagaraspl @lcobucci (and others ;)

DBAL https://github.com/doctrine/dbal/pull/2825 fixes the problem and should be ready to merge. Feel free to have a look (test or review)

*Till it's released I used to test my apps by updating my composer.json:*

```
    "repositories": [
        {
            "url": "https://github.com/doctrine/dbal.git",
            "type": "vcs"
        }
    ],
    "require": {
        "doctrine/dbal": "dev-fix/mariadb/2824 as 2.6.4",
```

*(It might not be the best way, and probably requires you add a token from github (when asked)... but helps to generate correct schema updates from my dev machine*

**Comment by lcobucci** [comment]
https://github.com/doctrine/dbal/pull/2825 has just been merged and it will be released on DBAL `v2.7.0`, please test if your issue is fixed when using `"doctrine/dbal": "2.7.x-dev@dev"` and let us know.

**Comment by Bukashk0zzz** [comment]
With "doctrine/dbal": "2.7.x-dev@dev" issue steel exist

**Comment by belgattitude** [comment]
@Bukashk0zzz 

Can you confirm, I don't have it anymore: My composer:

```
    "require": {
        "php": "^7.1",
        "doctrine/dbal": "2.7.x-dev as 2.6.4",
        "doctrine/common": "^2.7.0",
        "doctrine/orm": "^2.5.0",
        "gedmo/doctrine-extensions": "^2.4.28",
        "doctrine/data-fixtures": "^1.2.2",
        "symfony/console": "~2.6|~3.0|~4.0"
},
```

**Comment by Bukashk0zzz** [comment]
Tested with mariadb 10.2.11 and 10.3.1 official docker image.

in composer 
```json
        "symfony/orm-pack": "^1.0",
        "doctrine/dbal": "2.7.x-dev as 2.6.4",
```

Always get diff with null on every field.
On percona 5.7 and standard mysql 5.7 all works correct.

**Comment by belgattitude** [comment]
@Bukashk0zzz,

Have you recreated the database before testing ? It's possible you've already applied the migration with 'NULLs' prior to the patch (so your default values have been broken). Can be an possibility too

**Comment by Bukashk0zzz** [comment]
Made clear demo example https://github.com/Bukashk0zzz/dbal-test

With code in repo above I every time get diff. You can check it by yourself

<img width="1234" alt="screen shot 2017-12-04 at 12 57 10" src="https://user-images.githubusercontent.com/1908342/33549454-3589ce30-d8f3-11e7-88a3-6bf0c7f64bc0.png">

**Comment by Bukashk0zzz** [comment]
I have one entity https://github.com/Bukashk0zzz/dbal-test/blob/master/src/Entity/Admin.php

And one init migration https://github.com/Bukashk0zzz/dbal-test/blob/master/src/Migrations/Version20171204100535.php

How I run project https://github.com/Bukashk0zzz/dbal-test/blob/master/up.sh

**Comment by belgattitude** [comment]
@Bukashk0zzz 

Can you check by commenting `server_version: '5.7'` in `config/packages/doctrine.yaml` ?

Like this:
```
doctrine:
    dbal:
        # configure these for your database server
        driver: 'pdo_mysql'
        # server_version: '5.7'
        charset: utf8mb4

        # With Symfony 3.3, remove the `resolve:` prefix
        url: '%env(resolve:DATABASE_URL)%'

```

**Comment by belgattitude** [comment]
With the commented server_version, everything looks good so far. Am I missing something ?

![image](https://user-images.githubusercontent.com/259798/33550862-b4ea285e-d8ef-11e7-9e73-acae3671007c.png)

**Comment by Bukashk0zzz** [comment]
Still has problem with commented server_version: '5.7'


At your screenshot you just run migrations. Try run now for getting diff:
```bash
bin/console doctrine:migrations:diff
or
bin/console doctrine:schema:update --dump-sql
```

**Comment by belgattitude** [comment]
@Bukashk0zzz ...

Thanks, I can reproduce now... Could you help me to determine where the problem is ? I've just checked your [entity](https://github.com/Bukashk0zzz/dbal-test/blob/77ba5e40f1ccdc83ccdf67a6cd5c7e8e7ecc020a/src/Entity/Admin.php#L41), could you remove `unsigned` if you're using `booleans` ? 

```php
    /**
     * @var bool
     *
     * @ORM\Column(name="status", type="boolean", nullable=false, options={"default": true})
     */
    private $status = true;
 ```

And test with doctrine commands:

```shell
$ php bin/console doctrine:database:drop -n --force; php bin/console doctrine:database:create -n; php bin/console doctrine:schema:create -n
$ php bin/console doctrine:schema:update --dump-sql
```

It looks working... Either you change the `status` to `integer` or if you're using `boolean` remove `"unsigned"="true"`.

Should be the same using `migrations` commands too. 

Could you confirm you don't have migrations using mysql-5.7 ? So I can look deeper

**Comment by belgattitude** [comment]
Oups, just installed mysql-5.7... So with your entity, looks there's a diff generated as well:

```shell
$  rm -f src/Migrations/*.php
$ php bin/console doctrine:database:drop -n --force; php bin/console doctrine:database:create; php bin/console doctrine:schema:create -n
$ php bin/console doctrine:migrations:diff
# Generated new migration class to "/web/www/dbal-test/src/Migrations/Version20171204125419.php" from schema differences.
```

So I guess the same behaviour is for MySQL 5.7 too. So the way to fix is to fix your entity. Is there anything I missed ?

**Comment by Bukashk0zzz** [comment]
Updated repo. Now boolean not unsigned.

Still get diff:

 ALTER TABLE admins CHANGE login login VARCHAR(255) DEFAULT NULL, CHANGE password password VARCHAR(255) DEFAULT NULL;

Also on percona and official mysql all works correct.

**Comment by belgattitude** [comment]
@Bukashk0zzz 

Looks you've forget to remove 'server_version' ?  I've made a p/r, have a look:

Could you test by doing (exactly) this:

```php
$ composer update
$ rm -f src/Migrations/*.php
$ php bin/console doctrine:database:drop -n --force; php bin/console doctrine:database:create; php bin/console doctrine:schema:create -n
$ php bin/console doctrine:migrations:diff
```

What I have is 

```
# No changes detected in your mapping information.
```

**Comment by belgattitude** [comment]
Also note in your `symfony.lock`:

```php
    "doctrine/dbal": {
        "version": "v2.6.3"
    },
```

Not sure exactly how it can affect.... but it could be the problem. To be sure check:

```
$ ls -la vendor/doctrine/dbal/lib/Doctrine/DBAL/Platforms/Keywords
```

If you're on the good doctrine version, you should have a `MariaDb102Keywords.php` file in it.

**Comment by Bukashk0zzz** [comment]
Commenting 5.7 helped
Now all works correct. So there was two problems:
- # server_version: '5.7' in config
- unsigned bool in mapping

Add travis check for this case https://travis-ci.org/Bukashk0zzz/dbal-test/builds/311377116

**Comment by mdeboer** [comment]
Glad it's fixed (by the looks of it), when will this make it to master?

**Comment by belgattitude** [comment]
@Bukashk0zzz, happy for you. server_version will bypass database detection, that explains. For the other one: `unsigned bool in mapping`, unsigned boolean (really tinyint(4)) does not play well with schema manager (same applies for Mysql 5.7, could you confirm ?). 

@mdeboer Don't know when they are able to release, but in the meantime, please be welcome to test ;)

**Comment by Bukashk0zzz** [comment]
@belgattitude https://github.com/Bukashk0zzz/dbal-test/pull/2 as you can see mysql 5.7 have not any issue with unsigned int. There are problems only with mariaDB

Check travis buid https://travis-ci.org/Bukashk0zzz/dbal-test/builds/311772602?utm_source=github_status&utm_medium=notification

You can create pull-request to this repo and test any cases you want.

**Comment by belgattitude** [comment]
@Bukashk0zzz 

> as you can see mysql 5.7 have not any issue with unsigned bool

Nope ! Your travis build show there's no migration, right ! But your entity does not have `unsigned:true`. So it's meaningless...

To figure out, please test with 'unsigned:true' AND mysql5-7 (both conditions, don't forget to remove existing migrations `rm -f src/Migrations/*.php` if needed). 

Anyway I just did the test on my laptop with mysql-5.7. And I can confirm, there's a migration happing with unsigned bool on mysql5.7 too.

Thanks

**Comment by Bukashk0zzz** [comment]
@belgattitude You look at master not at PR and branch mysql

See PR files
https://github.com/Bukashk0zzz/dbal-test/pull/2/files

There are UNSIGNED in entity and in migration.

**Comment by belgattitude** [comment]
Ok, I still don't get it. Cannot reproduce but we are using 2 different ways to tests. I cannot help with `doctrine:migrations`... we need to start from dbal. Can you modify your travis scripts:

https://github.com/Bukashk0zzz/dbal-test/blob/b4dadcaa4672a314a40205368e798f523849b123/.travis.yml#L21

And only test with doctrine/dbal commands:

```
  - php bin/console doctrine:database:drop -n --force
  - php bin/console doctrine:database:create -n
  - php bin/console doctrine:schema:create -n
  - php bin/console doctrine:schema:update --dump-sql
  - php bin/console doctrine:schema:validate -n
```

This is what I have both unsigned:true and mysql 5.7

![image](https://user-images.githubusercontent.com/259798/33607484-931f2238-d9c1-11e7-8d03-90d249b1fb83.png)

I'm not using doctrine:migrations, that why I've always removed `rm -f src/Migrations/*.php` just to be sure we work on the same schema specs. If the test works above works for you, can you track it down to see where the migration process fails ?

**Comment by belgattitude** [comment]
For extended info, look what doctrine creates with `unsigned: bools` on mysql-5.7 or mariadb-10.2:

```
> php bin/console doctrine:schema:create --dump-sql
```

Here's the output (removed some columns for clarity):

```sql
CREATE TABLE admins (id INT UNSIGNED AUTO_INCREMENT NOT NULL,  status TINYINT(1) DEFAULT '1' NOT NULL, PRIMARY KEY(id))
```
It looks it didn't set the unsigned while creating... 

But from schema introspection (schema:update) it always trigger a diff if using unsigned. DBAL seems to alias BOOL to TINYINT(1) and seems to remove unsigned property (mysql or mariadb). I don't think https://github.com/doctrine/dbal/pull/2825/ have changed that.

My question now is more about how 'doctrine:migrations' handles that. Can you eventually see if there's differences in generated migrations for both databases ? It can help me track down the problem

Thanks

**Comment by belgattitude** [comment]
@Bukashk0zzz 

Maybe you can help me, is this how I can achieve a migration test from scratch with travis ?

```
  - php bin/console doctrine:database:drop -n --force
  - php bin/console doctrine:database:create -n
  - rm -f src/Migrations/*.php
  - php bin/console doctrine:schema:diff
  - php bin/console doctrine:schema:migrate -n
```

I would like to be sure there's no generated migration, that could potentially affect my tests?
I would like

**Comment by Bukashk0zzz** [comment]
@belgattitude You correct and on mysql and on mariadb with unsigned bool without my init migration always getting diff. Migrations generated by mysql and mariadb identical.

Answering on question how I get this migration:
I have already old project on ZF 1 and rewiring it to symfony 4 I just copy create table syntax from existing mysql db and set it to initial migration.

I agree with you that this not related to issue #2825

But what you think this are issue that dbal not correct handle unsigned bool ?
If yes I will create new issue. If no ok. Anyway I already fixed problem on my project.

travis test for mariadb https://travis-ci.org/Bukashk0zzz/dbal-test/builds/313452048

travis test for mysql https://travis-ci.org/Bukashk0zzz/dbal-test/builds/313450766

**Comment by mdeboer** [comment]
Why on earth would you use an unsigned bool anyway? Unless it's part of a bigger issue I don't see why this has priority to get fixed.

**Comment by Bukashk0zzz** [comment]
> Why on earth would you use an unsigned bool anyway?

But why not? Near 6-7 years ago I always put unsigned for every integer that will be only with +
And that was good practice. Now I am using doctrine for a while and don't really look what created in db because all this things hided.

Anyway I agree that this not an issue for now. Thats why I asked and not created issue for that.
Also I think this issue can be closed @belgattitude tested everything.

**Comment by lcobucci** [comment]
@Bukashk0zzz awesome, closing then!

**Comment by musicjerm** [comment]
This issue has resurfaced for me with a clean symfony/website-skeleton install.  Running MariaDB 10.2.14 and doctrine/dbal v2.7.0.  I compared against a fresh symfony-standard-edition connected to the same database.  Verified doctrine/* versions are the same.  Set up a simple entity to map with one field nullable=true and I get the issue on the symfony 4.0 install but not the 3.4.  Anyone else able to confirm my insanity?
